.github/workflows/teamcity-nightly-workflow.yaml

```diff
@@ -0,0 +1,83 @@
+                    owner: context.repo.owner,
+                    repo: context.repo.repo,
+                    ref: "heads/main"
+                  })
+                  const branchName = "UTC-nightly-tests-" + dateToday;
+                   const resp = await github.rest.git.createRef({
+                     owner: context.repo.owner,
+                     repo: context.repo.repo,
+                     ref: "refs/heads/" + branchName,
+                     sha: mainRef.data.object.sha
+                   })
+                   console.log("Created Branch: " + branchName)
+
+  branch-sweeper:
+    needs: create-branch
+    runs-on: ubuntu-latest
+    steps: 
+      - uses: actions/github-script@e69ef5462fd455e02edcaf4dd7708eda96b9eda0 # v7.0.0
+        env:
+          DAYS_THRESHOLD: ${{ inputs.dayThreshold }}
+        with:
+          script: |
+            const { DAYS_THRESHOLD } = process.env
+            console.log(`Removing nightly-test branches not made in the last ${DAYS_THRESHOLD} days`)
+        
+            function dateDifference(dateToday, branchDate){
+                dateToday = new Date(dateToday)
+                branchDate = new Date(branchDate)
+                return (dateToday - branchDate) / 86_400_000 // calculates the difference in days based on milliseconds
+            }
+            
+            async function branchSweeper(daysThreshold){
+                let dateToday = new Date().toJSON().slice(0, 10);
+                console.log("Today\'s date: ",dateToday);
+                // grab the list of branches then iterate through the list checking for the difference in days
+                const branchList = await github.rest.repos.listBranches({
+                    owner: context.repo.owner,
+                    repo: context.repo.repo,
+                  })
+                
+                for (let i = 0; i < branchList.data.length; i++) {
+                    const branchName = branchList.data.at(i).name
+                    const branchDate = /\d{4,4}-\d{1,2}-\d{1,2}/g.exec(branchName)
+                    if (branchDate == null) continue // skips if regex fails (is successful if matches with UTC-nightly-test branch format)
+                    if (dateDifference(dateToday, branchDate) >= daysThreshold) { // only happens if difference is greater than or equal to 3, we only want to keep the last 3 night branches
+                        const resp = await github.rest.git.deleteRef({
+                            owner: context.repo.owner,
+                            repo: context.repo.repo,
+                            ref: "heads/" + branchName,
+                          })
+                        if (resp.status == "422"){
+                            console.error("Branch doesn\'t exist")
+                        } else{
+                          console.log("Deleted branch: " + branchName)
+                        } 
+                    }
```

Comment:
Nice! I went with the first option of refining just regex since the resp doesn't given as an array of just the branch names it gives us an array of data from each branch.

So we would have to grab the resp, filter the resp to only have an array with the branch name's then filter using the `Array.filter`, and then go through the filtered array to check whether it's past the due date or not.

Let me know your thoughts, I can implement the second method if needed. Just sharing my thoughts 🤓

Reply:
The filter method can be used on an array containing any data type, if that helps with decision making?

E.g.
```js
const myObjects = [{ foo:'spraying'}, { foo:'elite'}];

const result = myObjects.filter((obj) => obj.foo.length > 6); // bear in mind there's "syntactic sugar" happening here

console.log(result);
```

So for the data returned about branches it could look like:

```js
const pretendDataFromApi = [{ name:'UTC-nightly-tests-2024-01-01'}, { name:'my-branch'}];

const nightlyTestBranches = pretendDataFromApi.filter( (x) => {
  const branchDate = /^UTC-nightly-tests-\d{4}-\d{2}-\d{2}$/g.exec(x.name)
  return branchDate != null // Need to return a boolean that's true for elements we want to keep
})
console.log("Found matching branches: ", nightlyTestBranches)
```

You could then filter the output  (i.e. nightlyTestBranches) again where the filter function pulls out the date from the name, returns a boolean reflecting if it's from over X days ago or not. And then you have a list of things to delete ⚔️

Reply:
that's awesome actually, i'll implement it using that instead!